### DISCUSSION TITLE
When can Relationship.local_remote_pairs be None?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#10161] [Usage Questions] [answered]
In 2.0.19 we get a type error because local_remote pairs says it can be `None` (In 2.0.18 it was still `Any`):
https://github.com/aio-libs/aiohttp-admin/actions/runs/5701188662/job/15451661240?pr=752

Under what condition can this be `None`?

**Comment by zzzeek** [comment]
currently, when you first construct a `relationship()`, that attribute is None.  It is immediately assigned the value of the parameter `_local_remote_pairs` which was provided when the "local remote" refactor first occurred many years ago, as a way to work around issues that might have been present for some complex join conditions.  This parameter defaults to `None`.

once the `configure_mappers()` operation occurs, `relationship()` figures out the effective value of `local_remote_pairs` and then it's set up.

**Reply by Dreamsorcerer** [reply, marked as answer]
OK, so if I'm parsing an ORM model with relationships on it, it should never be `None`? So, I can just add an `assert` for type checking.

**Reply by zzzeek** [reply]
yes you can assert it
